File: ai_router/modules.py
"""Module manifest: the one declaration of what a module is.

The manifest is a YAML mapping with a ``modules`` list; each entry is
``{slug, title?, planPath?, codeRoots?, touches?, specSections?,
contextAssets?}``. ``codeRoots`` bounds the module on disk,
``specSections`` maps reference spec sections to it, and
``contextAssets`` names its schemas/config/migrations. The extension's
reader takes the keys it knows and ignores the rest, so the extension
keeps rendering entries carrying the newer keys.

An unknown key is rejected rather than ignored: a misspelled
``codeRoot`` that is silently dropped leaves the module bounded by
something other than what was written, which is the failure this
manifest exists to prevent.

Create-only by design: rename, delete, and reorganization stay manual
edits to the file.
"""

from __future__ import annotations

import argparse
import json
import sys
from dataclasses import dataclass
from pathlib import Path
from typing import Optional

import yaml

EXIT_OK = 0
EXIT_REFUSED = 1
EXIT_USAGE = 2

MANIFEST_RELPATH = Path("docs") / "modules.yaml"

KNOWN_ENTRY_KEYS = (
    "slug", "title", "planPath", "codeRoots", "touches", "specSections",
    "contextAssets",
)

_LIST_KEYS = ("codeRoots", "touches", "specSections", "contextAssets")


@dataclass(frozen=True)
class ModuleEntry:
    """One validated manifest entry, in Python spelling."""

    slug: str
    title: str
    plan_path: Optional[str] = None
    code_roots: tuple = ()
    touches: tuple = ()
    spec_sections: tuple = ()
    context_assets: tuple = ()


def manifest_path(workspace_root) -> Path:
    return Path(workspace_root) / MANIFEST_RELPATH


def load_manifest(path: Path) -> dict:
    """The parsed manifest mapping. A missing file is the designed empty
    state (``{"modules": []}``); a bare ``modules:`` (YAML null) is a valid
    empty list. Unparseable YAML, a non-mapping document, or a ``modules``
    value that is neither null nor a list raises ``ValueError`` — a config
    error must refuse loud, never be silently rewritten."""
    if not path.is_file():
        return {"modules": []}
    try:
        doc = yaml.safe_load(path.read_text(encoding="utf-8"))
    except yaml.YAMLError as exc:
        raise ValueError(f"{path} is not valid YAML: {exc}") from exc
    if doc is None:
        doc = {}
    if not isinstance(doc, dict):
        raise ValueError(
            f"{path} must be a YAML mapping with a 'modules' list"
        )
    modules = doc.get("modules")
    if modules is None:
        doc["modules"] = []
    elif not isinstance(modules, list):
        raise ValueError(f"'modules' in {path} must be a list")
    return doc


def _string_list(value, where: str, key: str) -> tuple:
    if value is None:
        return ()
    if not isinstance(value, list):
        raise ValueError(f"{where}: '{key}' must be a list of strings")
    out = []
    for item in value:
        if not isinstance(item, str) or not item.strip():
            raise ValueError(
                f"{where}: '{key}' must contain only non-empty strings"
            )
        out.append(item.strip())
    return tuple(out)


def parse_entries(doc: dict, source="docs/modules.yaml") -> list:
    """Validated entries in file order. Rejects an unknown key, a
    non-mapping entry, a missing slug, a duplicate slug, and a
    mistyped list — never silently drops one."""
    entries: list = []
    seen: set = set()
    for index, raw in enumerate(doc.get("modules") or []):
        where = f"{source}: modules[{index}]"
        if not isinstance(raw, dict):
            raise ValueError(f"{where} must be a mapping")
        unknown = sorted(set(raw) - set(KNOWN_ENTRY_KEYS))
        if unknown:
            raise ValueError(
                f"{where} has unknown key(s) {', '.join(unknown)}. "
                f"Known keys: {', '.join(KNOWN_ENTRY_KEYS)}."
            )
        slug = raw.get("slug")
        if not isinstance(slug, str) or not slug.strip():
            raise ValueError(f"{where} needs a non-empty string 'slug'")
        slug = slug.strip()
        if slug in seen:
            raise ValueError(f"{source}: duplicate slug {slug!r}")
        seen.add(slug)
        title = raw.get("title")
        if title is not None and not isinstance(title, str):
            raise ValueError(f"{where}: 'title' must be a string")
        plan_path = raw.get("planPath")
        if plan_path is not None and not isinstance(plan_path, str):
            raise ValueError(f"{where}: 'planPath' must be a string")
        lists = {
            key: _string_list(raw.get(key), where, key) for key in _LIST_KEYS
        }
        entries.append(ModuleEntry(
            slug=slug,
            title=(title or "").strip() or slug,
            plan_path=(plan_path or "").strip() or None,
            code_roots=lists["codeRoots"],
            touches=lists["touches"],
            spec_sections=lists["specSections"],
            context_assets=lists["contextAssets"],
        ))
    return entries


def load_entries(workspace_root) -> list:
    """Validated entries for a workspace. An absent manifest is the
    designed empty state; an invalid one raises ``ValueError``."""
    path = manifest_path(workspace_root)
    return parse_entries(load_manifest(path), source=str(path))


def find_entry(workspace_root, slug: str):
    """The entry for *slug*, or ``None`` when the manifest does not
    declare it — an unresolvable slug is the caller's cue to fall back,
    never to guess at what the module covers."""
    if not slug:
        return None
    wanted = slug.strip()
    for entry in load_entries(workspace_root):
        if entry.slug == wanted:
            return entry
    return None


def create(workspace_root, slug: str, title: str, plan_path=None,
           code_roots=None, spec_sections=None, context_assets=None) -> int:
    path = manifest_path(workspace_root)
    try:
        doc = load_manifest(path)
        parse_entries(doc, source=str(path))
    except ValueError as exc:
        print(f"modules create: refused -- {exc}", file=sys.stderr)
        return EXIT_REFUSED
    existing = {
        entry.get("slug")
        for entry in doc["modules"] if isinstance(entry, dict)
    }
    if slug in existing:
        print(
            f"modules create: refused -- slug {slug!r} already exists "
            f"in {path}", file=sys.stderr,
        )
        return EXIT_REFUSED
    entry = {"slug": slug, "title": title}
    if plan_path:
        entry["planPath"] = plan_path
    for key, values in (
        ("codeRoots", code_roots),
        ("specSections", spec_sections),
        ("contextAssets", context_assets),
    ):
        if values:
            entry[key] = list(values)
    doc["modules"].append(entry)
    try:
        parse_entries(doc, source=str(path))
    except ValueError as exc:
        print(f"modules create: refused -- {exc}", file=sys.stderr)
        return EXIT_REFUSED
    path.parent.mkdir(parents=True, exist_ok=True)
    with open(path, "w", encoding="utf-8", newline="") as f:
        yaml.safe_dump(doc, f, sort_keys=False, allow_unicode=True,
                       default_flow_style=False)
    print(json.dumps(entry))
    return EXIT_OK


def main(argv=None) -> int:
    parser = argparse.ArgumentParser(prog="python -m ai_router.modules")
    sub = parser.add_subparsers(dest="command", required=True)
    p_create = sub.add_parser(
        "create", help="append one module entry to docs/modules.yaml"
    )
    p_create.add_argument("workspace_root",
                          help="workspace root containing docs/")
    p_create.add_argument("--slug", required=True,
                          help="machine identity (kebab-case)")
    p_create.add_argument("--title", required=True,
                          help="display name the Explorer shows")
    p_create.add_argument("--plan-path",
                          help="module plan path, relative to the root")
    p_create.add_argument("--code-root", action="append", dest="code_roots",
                          help="repo-relative directory that bounds the "
                               "module on disk (repeatable)")
    p_create.add_argument("--spec-section", action="append",
                          dest="spec_sections",
                          help="reference spec section as PATH or "
                               "PATH#anchor (repeatable)")
    p_create.add_argument("--context-asset", action="append",
                          dest="context_assets",
                          help="schema/config/migration path or glob "
                               "(repeatable)")
    args = parser.parse_args(argv)
    if not Path(args.workspace_root).is_dir():
        print(f"modules: not a directory: {args.workspace_root}",
              file=sys.stderr)
        return EXIT_USAGE
    return create(args.workspace_root, args.slug, args.title,
                  args.plan_path, args.code_roots, args.spec_sections,
                  args.context_assets)


if __name__ == "__main__":
    raise SystemExit(main())

File: packages/router/src/cli/modules.ts
// `dabbler modules create` -- append one entry to `docs/modules.yaml`;
// `dabbler modules show` -- print the manifest as the framework reads it.
//
// One writer, because the manifest has one writer. The `Router` contract
// used to name `list` and `retire`; nothing implemented either, and a
// contract naming a verb nothing implements is a promise that would be
// refused at the moment it was needed (D162/D152). `show` is the reader:
// the shape, the entries in dependency order, and `usedBy` derived. Rename,
// delete and reorganization stay manual edits to the manifest.

import { EXIT_OK as CREATED, create, show } from "../modules.ts";
import { tryWriteProjection } from "../projection.ts";
import { writeErr, writeOut } from "./output.ts";
import { statSync } from "node:fs";

const EXIT_OK = 0;
const EXIT_USAGE = 2;

/** Flags that may be given more than once, each occurrence appending. */
const REPEATABLE: Readonly<Record<string, string>> = {
  "--code-root": "codeRoots",
  "--depends-on": "dependsOn",
  "--spec-section": "specSections",
  "--context-asset": "contextAssets",
};

const SINGLE = new Set([
  "--slug",
  "--title",
  "--plan-path",
  "--kind",
  "--package",
  "--contract",
]);

function usage(): string {
  return [
    "usage: dabbler modules create [-h] --slug SLUG --title TITLE",
    "                              [--plan-path PLAN_PATH]",
    "                              [--code-root CODE_ROOTS]",
    "                              [--kind {shared-types,library,application}]",
    "                              [--depends-on SLUG]",
    "                              [--package PACKAGE]",
    "                              [--contract {designed,package,generated}]",
    "                              [--spec-section SPEC_SECTIONS]",
    "                              [--context-asset CONTEXT_ASSETS]",
    "                              workspace_root",
    "       dabbler modules show [-h] workspace_root",
    "",
    "positional arguments:",
    "  workspace_root        workspace root containing docs/",
    "",
    "options:",
    "  -h, --help            show this help message and exit",
    "  --slug SLUG           machine identity (kebab-case)",
    "  --title TITLE         display name the Explorer shows",
    "  --plan-path PLAN_PATH",
    "                        module plan path, relative to the root",
    "  --code-root CODE_ROOTS",
    "                        repo-relative directory that bounds the module on",
    "                        disk (repeatable)",
    "  --kind KIND           shared-types, library (the default) or application",
    "  --depends-on SLUG     a module this one consumes (repeatable); who",
    "                        depends on a module is derived, never declared",
    "  --package PACKAGE     the artifact id a sibling consumes (a NuGet id, or",
    "                        Maven's groupId:artifactId)",
    "  --contract MODE       designed, package (the default when a package is",
    "                        declared) or generated",
    "  --spec-section SPEC_SECTIONS",
    "                        reference spec section as PATH or PATH#anchor",
    "                        (repeatable)",
    "  --context-asset CONTEXT_ASSETS",
    "                        schema/config/migration path or glob (repeatable)",
    "",
    "`show` prints the manifest as the framework reads it: the shape (one",
    "module, or many), the entries in dependency order, and usedBy per module.",
    "",
  ].join("\n");
}

function isDirectory(path: string): boolean {
  try {
    return statSync(path).isDirectory();
  } catch {
    return false;
  }
}

export async function modulesVerb(argv: string[]): Promise<number> {
  const [subcommand, ...rest] = argv;
  if (subcommand === undefined) {
    writeErr(
      `dabbler modules: the following arguments are required: command\n\n${usage()}`,
    );
    return EXIT_USAGE;
  }
  if (subcommand === "--help" || subcommand === "-h") {
    writeOut(usage());
    return EXIT_OK;
  }
  if (subcommand === "show") {
    const [root, ...extra] = rest;
    if (root === "--help" || root === "-h") {
      writeOut(usage());
      return EXIT_OK;
    }
    if (root === undefined || extra.length > 0) {
      writeErr(`dabbler modules show: the following arguments are required: workspace_root\n`);
      return EXIT_USAGE;
    }
    if (!isDirectory(root)) {
      writeErr(`modules: not a directory: ${root}\n`);
      return EXIT_USAGE;
    }
    return show(root);
  }
  if (subcommand !== "create") {
    writeErr(`dabbler modules: '${subcommand}' is not a subcommand\n\n${usage()}`);
    return EXIT_USAGE;
  }

  const single = new Map<string, string>();
  const repeated = new Map<string, string[]>();
  const positional: string[] = [];
  for (let index = 0; index < rest.length; index += 1) {
    const token = rest[index];
    if (token === "--help" || token === "-h") {
      writeOut(usage());
      return EXIT_OK;
    }
    if (!token.startsWith("--")) {
      positional.push(token);
      continue;
    }
    const equals = token.indexOf("=");
    const flag = equals === -1 ? token : token.slice(0, equals);
    if (!SINGLE.has(flag) && !(flag in REPEATABLE)) {
      writeErr(`dabbler modules create: unrecognized argument: ${flag}\n\n${usage()}`);
      return EXIT_USAGE;
    }
    let value: string;
    if (equals !== -1) {
      value = token.slice(equals + 1);
    } else {
      const next = rest[index + 1];
      if (next === undefined || next.startsWith("--")) {
        writeErr(`dabbler modules create: argument ${flag}: expected one argument\n`);
        return EXIT_USAGE;
      }
      value = next;
      index += 1;
    }
    if (flag in REPEATABLE) {
      const key = REPEATABLE[flag];
      repeated.set(key, [...(repeated.get(key) ?? []), value]);
    } else {
      single.set(flag, value);
    }
  }

  const missing = ["--slug", "--title"].filter((flag) => !single.has(flag));
  if (positional.length === 0) missing.unshift("workspace_root");
  if (missing.length > 0) {
    writeErr(
      `dabbler modules create: the following arguments are required: ${missing.join(", ")}\n`,
    );
    return EXIT_USAGE;
  }
  const workspaceRoot = positional[0];
  if (!isDirectory(workspaceRoot)) {
    writeErr(`modules: not a directory: ${workspaceRoot}\n`);
    return EXIT_USAGE;
  }
  const code = create(workspaceRoot, single.get("--slug")!, single.get("--title")!, {
    planPath: single.get("--plan-path") ?? null,
    codeRoots: repeated.get("codeRoots") ?? null,
    specSections: repeated.get("specSections") ?? null,
    contextAssets: repeated.get("contextAssets") ?? null,
    kind: single.get("--kind") ?? null,
    dependsOn: repeated.get("dependsOn") ?? null,
    package: single.get("--package") ?? null,
    contract: single.get("--contract") ?? null,
  });
  // The manifest moved, so the projection the Solution Explorer reads is
  // rewritten here, by the verb that moved it -- the same rule every other
  // declaration-moving verb follows. The extension also re-derives on a
  // manifest change it watches, but a verb run from a terminal has no
  // extension to do it for it.
  if (code === CREATED) tryWriteProjection(workspaceRoot);
  return code;
}
